src/node/redeem.rs

```diff
@@ -273,6 +274,162 @@ impl<J: Jet> RedeemNode<J> {
+        }
+
+        impl<J: Jet> Converter<Redeem<J>, Witness<J>> for Pruner<J> {
+            type Error = std::convert::Infallible;
+
+            fn convert_witness(
+                &mut self,
+                data: &PostOrderIterItem<&RedeemNode<J>>,
+                witness: &Value,
+            ) -> Result<Option<Value>, Self::Error> {
+                let pruned = witness.prune(&data.node.arrow().target);
+                debug_assert!(
+                    pruned.is_some(),
+                    "pruned type should be shrunken version of unpruned type"
+                );
+                Ok(pruned)
+            }
+
+            fn convert_disconnect(
+                &mut self,
+                _: &PostOrderIterItem<&RedeemNode<J>>,
+                right: Option<&Arc<WitnessNode<J>>>,
+                _: &Arc<RedeemNode<J>>,
+            ) -> Result<Option<Arc<WitnessNode<J>>>, Self::Error> {
+                debug_assert!(
+                    right.is_some(),
+                    "disconnected branch should exist in unpruned redeem program"
+                );
+                Ok(right.map(Arc::clone))
+            }
+
+            fn prune_case(
+                &mut self,
+                data: &PostOrderIterItem<&RedeemNode<J>>,
+                _left: &Arc<WitnessNode<J>>,
+                _right: &Arc<WitnessNode<J>>,
+            ) -> Result<Hide, Self::Error> {
+                // The IMR of the pruned program may change,
+                // but the Converter trait gives us access to the unpruned node (`data`).
+                // The Bit Machine tracked (un)used case branches based on the unpruned IMR.
+                match (
+                    self.executed_left.contains(&data.node.imr()),
+                    self.executed_right.contains(&data.node.imr()),
+                ) {
+                    (true, true) => Ok(Hide::Neither),
+                    (false, true) => Ok(Hide::Left),
+                    (true, false) => Ok(Hide::Right),
+                    (false, false) => Ok(Hide::Neither), // case nodes that were never executed will be pruned out by their ancestors
+                }
+            }
+
+            fn convert_data(
+                &mut self,
+                _: &PostOrderIterItem<&RedeemNode<J>>,
+                inner: Inner<&Arc<WitnessNode<J>>, J, &Option<Arc<WitnessNode<J>>>, &Option<Value>>,
+            ) -> Result<WitnessData<J>, Self::Error> {
+                let converted_inner = inner
+                    .map(|node| node.cached_data())
+                    .map_witness(Option::<Value>::clone);
+                let retyped = WitnessData::from_inner(&self.inference_context, converted_inner)
+                    .expect("pruned types should check out if unpruned types check out");
+                Ok(retyped)
+            }
+        }
+
+        struct Finalizer<J>(PhantomData<J>);
+
+        impl<J: Jet> Converter<Witness<J>, Redeem<J>> for Finalizer<J> {
```

Comment:
In 7d93c829eed0ca9935c595214c5fa526564f4bf8:

This maybe should also be a followup PR, because I'm unsure of correctness, but are you sure we need to do this in two complete passes? Or could we do combine the pruner and finalizer into one?

Even if the answer is no it might be worthwhile to try it, fuzz for programs that pass on one but not the other, and use them as unit tests..

Reply:
I think there are non-local effects in the type inference. We convert the program in post-order, so a witness node etc. cannot know the pruned type that it will receive. _This also means that I implemented the pruning of witness values incorrectly, because I used the unpruned type. I will fix this and extend the unit tests._